sources/typescript/src/adapters/kaltura-upload-connections-manager.ts

```diff
@@ -0,0 +1,71 @@
+export class KalturaUploadConnectionsManager {
+    private static _availableConnections: number;
+    private static _callbacks: (() => void)[];
+    private static _totalConnections: number = -1; // -1 as in not initialized
+    private static _initialized = false;
+
+    /**
+     * initialize the manager. will only do something if not yet initialized.
+     * @param maxConnections  max number of concurrent connections
+     */
+    public static init(maxConnections: number) {
+        // only if not yet initialized
+        if (!this._initialized) {
+            this._totalConnections = maxConnections;
+            this._availableConnections = maxConnections;
+            // this is a fine opportunity to initialize this as well:
+            this._callbacks = [];
+            this._initialized = true;
+        }
+    }
+
+    /**
+     * let an adapter use a connection
+     * @return true if there is an available connection, false otherwise
+     */
+    public static retrieveConnection(): boolean {
+        if (this._availableConnections >= 1) {
+            this._availableConnections -= 1;
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * return a connection to the pool
+     */
+    public static releaseConnection(): void {
+        this._availableConnections += 1;
```

Comment:
worth checking that there's no case where available connections > totalConnections.
Looking at the current code, the total is only set during init, but there's no verification that we're not passing it.
in fact, currently totalConnections is not used - so if it is not needed, better remove it from the code.

Reply:
I will keep it and add a log if we are passing it, which I hope we will never see..

Reply:
client side log - surely enough if it happens to a customer we won't see it...
why not add a condition (that might never be met) like:
```
if(this._availableConnections + 1 <= this._totalConnections) {
  this._availableConnections +=1;
}
```
?